### ISSUE TITLE
To use the amg(amg.hpp) as a standalone solver.

### ISSUE THREAD

**Issue by g873249824** [ddemidov/amgcl#172]
I want to use the amg as a stand alone solver for some tests with changes to the algorithm to accelerate the convergence,
what should i pay attention with the code? maybe i can refer to the examples in /solver (cg gmres e.t.c) to keep cycling(cycle()) until it satisfies the break condition.

**Comment by ddemidov** [comment]
You are correct, you only need to instantiate `amgcl::amg<...>` and keep [cycling](https://github.com/ddemidov/amgcl/blob/affc72ec009393134688e17110a8d38661d2a99a/amgcl/amg.hpp#L221-L229) until convergence. Something like this should work:
```cpp
amgcl::amg<...> amg(A, prm);
while(true) {
    amg.cycle(f, x);
    amgcl::backend::residual(f, A, x, res);
    if (norm(res) / norm(f) < tol) break;
}
```

**Comment by g873249824** [comment]
pergectly!Thank you, I got the confidence from your reply. I was trying to overload the operator () to make it.

**Comment by g873249824** [comment]
I have another questions:
1. If i get the matrix R(n,m) V(m,n), the row of R or the colum of V are the vector of residual or something like that(n x 1), and m is like 10 or 8 or 7. I want to calculte the  pinv(RV) . if i can use some functions implemented in /backend , which i mean i can combine(r1, r2......rm) as matrix and apply matrix to matrix mutiply and inverse. Maybe i think i need to get inner_product to get (m x m) elements and then get the pinv(RV).

**Comment by ddemidov** [comment]
Sorry, can you rephrase the question?

**Comment by g873249824** [comment]
Supposing the matrix R(n,m) V(m,n) are the combination of m history residul vectors or approximate vectors. I want to caculate the pseudoinverse of M=pseudoinverse(R*V) . N=M*V  .or Q=R*N.  how to implenment it efficiently.

**Comment by ddemidov** [comment]
amgcl will not help you in computing a pseudoinverse of a dense matrix, there is no functionality for that in the library.

**Comment by g873249824** [comment]
thank you, 
1.does /amgcl/detail/inverse only work for sparse matrix?
2.do i need to implement the dense matrix multiplication, or the spmv is a alternative way for that?

**Comment by ddemidov** [comment]
> does /amgcl/detail/inverse only work for sparse matrix?

amgcl/detail/inverse works with dense matrices, but first, it is `detail` namespace, which means it should not be used in user code, and second, it is far from effective (it is used internally for inverting small statically sized matrices, something like 2x2 or 3x3).

> do i need to implement the dense matrix multiplication, or the spmv is a alternative way for that?

I think you would be better off by using a library that was specifically designed for efficient dense linear algebra operations, such as a BLAS implementation.

**Comment by g873249824** [comment]
I got that.Thanks for your patience.

**Comment by ddemidov** [comment]
A classical example would be an [OpenBLAS](https://www.openblas.net/).
[Eigen](http://eigen.tuxfamily.org/) or [Blaze](https://bitbucket.org/blaze-lib/blaze/src/master/) have C++ interface and should be fast as well.

**Comment by g873249824** [comment]
Thank you，I got that.
